project

thegreatestacc opened the pull request.

sorryimbrainlessdeveloper committed:
project

AlohaJava commented on src/main/java/org/sovliv/kafkatask/KafkaTaskApplication.java:
```diff
@@ -0,0 +1,15 @@
+import org.springframework.kafka.annotation.EnableKafka;
+
+@SpringBootApplication
+@EnableKafka
```
лишнее

AlohaJava commented on src/main/java/org/sovliv/kafkatask/config/KafkaTopicConfig.java:
```diff
@@ -0,0 +1,27 @@
+@Configuration
+public class KafkaTopicConfig {
+
+    @Bean
```
можно без этого, строкой в kafkalistener
избыточно

AlohaJava commented on src/main/java/org/sovliv/kafkatask/controller/ProduceController.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @GetMapping("/ping")
+    public ResponseEntity<String> ping() {
```
actuator

AlohaJava replied:
/actuator/health

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/DataBaseToKafkaService.java:
```diff
@@ -0,0 +1,31 @@
+    private final KafkaTemplate<UUID, String> kafkaTemplate;
+
+    public void publishAllToKafka() {
+        repository.findAll().forEach(entity ->
```
не, 20 млн выгрузить так у меня тачка на 1 гб вылетит

AlohaJava replied:
давай пачками и на пуле

AlohaJava replied:
да го пачками
limit offset

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/DataBaseWriterHiber.java:
```diff
@@ -0,0 +1,81 @@
+    @PostConstruct
+    public void startDbWriters() {
+        for (int i = 0; i < WORKER_THREADS; i++) {
+            executorService.submit(() -> {
```
слишком сложно, разнеси на методы

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/DataBaseWriterJDBC.java:
```diff
@@ -0,0 +1,100 @@
+                        if (data != null) buffer.add(data);
+
+                        if (buffer.size() >= BATCH_SIZE || (!running && !buffer.isEmpty())) {
+                            saveBuffer(new ArrayList<>(buffer));
```
транзакшионал из паблик внутри

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/DataBaseWriterJDBC.java:
```diff
@@ -0,0 +1,100 @@
+    @Transactional
+    public void saveBuffer(List<SomeData> bufferToWrite) {
+        if (bufferToWrite.isEmpty()) return;
+        String sql = "INSERT INTO some_data (id, value) VALUES (?, ?) ON CONFLICT (id) DO NOTHING";
```
в константы

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/ProducerService.java:
```diff
@@ -0,0 +1,181 @@
+    public void sendMessages(int totalCount) {
+        System.out.println(String.format(STARTING_TO_SEND_MESSAGES, totalCount));
+
+        ExecutorService executor = Executors.newVirtualThreadPerTaskExecutor();
```
красава

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/ProducerService.java:
```diff
@@ -0,0 +1,181 @@
+
+        try {
+            latch.await();
+            System.out.println(ALL_MESSAGES_SENT);
```
на логгер го заменим

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/TpsLimiter.java:
```diff
@@ -0,0 +1,34 @@
+
+        // Обновляет лимит каждую секунду
+        Executors.newSingleThreadScheduledExecutor().scheduleAtFixedRate(() -> {
+            int toRelease = this.tps - permits.availablePermits();
```
чет жесть как сложно

AlohaJava replied:
давай попробуем проще средствами executors

AlohaJava commented on docker-compose.yml:
```diff
@@ -0,0 +1,128 @@
+    volumes:
+      - ./db-init:/docker-entrypoint-initdb.d
+
+    #    volumes:
```
а чё закомментил

AlohaJava commented on format-kafka.ps1:
```diff
@@ -0,0 +1,68 @@
+$clusterId = "xe2d5Se9STyfJOLtJGruMQ"
+$image = "confluentinc/cp-kafka:7.5.0"
+
+Write-Host "Formatting Kafka brokers with KRaft Cluster ID: $clusterId"
```
что-то слишком сложнаа

AlohaJava commented on kafka-cluster/kafka-1/config/server.properties:
```diff
@@ -0,0 +1,26 @@
+listener.security.protocol.map=PLAINTEXT:PLAINTEXT,CONTROLLER:PLAINTEXT
+log.dirs=/var/lib/kafka/data
+
+# Performance tuning
```
походу они с env дублируются

AlohaJava commented on pom.xml:
```diff
@@ -0,0 +1,110 @@
+        <url/>
+    </scm>
+    <properties>
+        <java.version>22</java.version>
```
а что не 24?

AlohaJava commented on resources/k1.conf:
```diff
@@ -0,0 +1,7 @@
+﻿process.roles=broker,controller
```
чё это за конфиг кафка кластера в ресурсах?

AlohaJava commented on resources/kafka-1.properties:
```diff
@@ -0,0 +1,7 @@
+﻿process.roles=broker,controller
```
тож не пон

AlohaJava commented on src/main/java/org/sovliv/kafkatask/controller/ProduceController.java:
```diff
@@ -0,0 +1,34 @@
+
+    @PostMapping
+    public ResponseEntity<String> produce(@RequestParam(defaultValue = "20000000") int count) {
+        producerService.sendMessages(count);
```
можно сделать асинхронной ручкой

AlohaJava replied:
чому не сделал

AlohaJava commented on src/main/java/org/sovliv/kafkatask/controller/ProduceController.java:
```diff
@@ -0,0 +1,34 @@
+    }
+
+    @GetMapping("/ping")
+    public ResponseEntity<String> ping() {
```
зачем обернул в ResponseEntity

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/DataBaseWriter.java:
```diff
@@ -0,0 +1,14 @@
+
+public interface DataBaseWriter {
+
+    boolean isDbOverloaded();
```
давай без интерфейсов с 1 реализацией

sorryimbrainlessdeveloper committed:
reworked writing implementation

AlohaJava commented on src/main/java/org/sovliv/kafkatask/mapper/SomeDataMapper.java:
```diff
@@ -0,0 +1,21 @@
+public class SomeDataMapper {
+
+    public SomeData toEntity(SomeDataDTO dto) {
+        return new SomeData(UUID.randomUUID(), dto.getValue());
```
go mapstruct

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/DBWriter.java:
```diff
@@ -0,0 +1,71 @@
+        if (buffer.isEmpty()) return;
+
+        try (Connection conn = dataSource.getConnection()) {
+            CopyManager copyManager = new CopyManager(conn.unwrap(BaseConnection.class));
```
жесть, ты это имел ввиду, не давай на jdbctemplate, там есть batch

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/ProducerService.java:
```diff
@@ -0,0 +1,107 @@
+                String json = mapper.writeValueAsString(someDataDTO);
+
+                CompletableFuture<SendResult<UUID, String>> future =
+                        kafkaTemplate.send(DATA_TOPIC, json).toCompletableFuture();
```
лучше це переписать, выглядит страшна
или можем рассмотреть уже на уроке

AlohaJava commented on src/main/java/org/sovliv/kafkatask/service/implv1_deprecated/BufferManager.java:
```diff
@@ -0,0 +1,70 @@
+    }
+
+    private synchronized void rotateQueue() {
+        BlockingQueue<SomeData> fullQueue = currentQueue;
```
избавиться бы от него, оверинженеринг

sorryimbrainlessdeveloper committed:
пишу в staging таблицу.
пишется оч быстро, но мне кажется надо писать в обычную таблицу. а так как сделано сейчас - неверно

sorryimbrainlessdeveloper committed:
reworked db writer

sorryimbrainlessdeveloper committed:
work in progress

sorryimbrainlessdeveloper committed:
fix script
